flow/environments/incite.py

```diff
@@ -207,4 +208,56 @@ def add_args(cls, parser):
         )
 
 
-__all__ = ["SummitEnvironment", "AndesEnvironment"]
+class CrusherEnvironment(DefaultSlurmEnvironment):
+    """Environment profile for the Cluster supercomputer.
+
+    https://docs.olcf.ornl.gov/systems/crusher_quick_start_guide.html
+    """
+
+    hostname_pattern = r".*\.crusher\.olcf\.ornl\.gov"
+    template = "crusher.sh"
+    cores_per_node = 128
+    gpus_per_node = 8
+    mpi_cmd = "srun"
+
+    @template_filter
+    def calc_num_nodes(cls, ngpus, ncpus, threshold):
+        """Compute the number of nodes needed to meet the resource request.
+
+        Also raise an error when the requested resource do not come close to saturating the asked
+        for nodes.
+        """
+        nodes_gpu = max(1, int(ceil(ngpus / cls.gpus_per_node)))
+        nodes_cpu = max(1, int(ceil(ngpus / cls.gpus_per_node)))
+        if nodes_gpu >= nodes_cpu:
+            check_utilization(nodes_gpu, ngpus, 8, threshold, "compute")
+            return nodes_gpu
+        check_utilization(nodes_cpu, ncpus, 64, threshold, "compute")
+        return nodes_cpu
+
+    @classmethod
+    def _get_mpi_prefix(cls, operation, parallel):
+        """Get the correct srun command for the job.
+
+        We don't currently support CPU/GPU mapping and expect the program to do this in code.
+        """
+        ngpus = operation.directives["ngpu"]
+        nranks = operation.directives.get("nranks", 1)
+        ncpus = max(
+            nranks * operation.directives.get("omp_num_threads", 1),
+            operation.directives.get("np", 1),
+        )
```

Comment:
Great questions. I think there's certainly room for improvement. I'll try to answer these based on my knowledge of project history:

If not provided, `np` is the product of `nranks` and `omp_num_threads` (threads per rank). Typically only `np` OR (`nranks`, `omp_num_threads`) should be specified -- but there are exceptions. There are use cases for `np` only without `nranks`, like Python multiprocessing which shouldn't use MPI (which specifying `nranks` always assumes). Likewise, there are cases for `np` and `nranks` if threading is handled some way besides OpenMP (because specifying `omp_num_threads` exports an environment variable).

I agree "task" could be clarified in the docs. Yes, `omp_num_threads` is threads per rank. All templates are supposed to handle these parameters the same way (not sure if that's presently true, but it should be).

Reply:
Bradley's explanation covers most of the main points. As to whether different templates handle these differently, I think `omp_num_threads` as a per-rank setting is fairly consistent. However, there are other possible differences between clusters. I don't recall all of the details any more, but as Bradley points out, `np` and `nranks` are not always the same given the use of e.g. Python multiprocessing, and in some cases I think those are cluster-specific differences (e.g. Stampede's launcher). In other cases, there is some subtle interplay between the parameters to support nontrivial translations to cluster-specific resource specifications (e.g. Summit's resource sets). I am sure that we could improve the current setup, but unfortunately simply removing parameters (for instance, removing either np or nranks) would probably shrink the set of cases that we currently support.

Reply:
Thanks for clarifying. I suppose such issues are unavoidable when mapping any form of generic resource definition to a system specific one. For example, another area I have trouble with flow is the how the `memory` request is passed to SLURM. It sets the **per-node** memory in the sbatch script. However, at the time the resource definition is written, the author does not know the number of nodes the job will eventually run on, so there is no way to compute a cross-platform memory request. I am working around this by setting `SBATCH` environment variables before submitting: e.g.
```
SBATCH_MEM_PER_CPU="4g" python hoomd_validation/project.py submit -o '.*_cpu'
```